pkg/disk/partition_table.go

```diff
@@ -160,6 +161,146 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	PartitionTableType PTType
+
+	// BootMode determines the types of boot-related partitions that are
+	// automatically added, BIOS boot (legacy), ESP (UEFI), or both (hybrid).
+	// If none, no boot-related partitions are created.
+	BootMode platform.BootMode
+
+	// DefaultFSType determines the filesystem type for automatically created
+	// filesystems and custom mountpoints that don't specify a type.
+	// None is only valid if no partitions are created and all mountpoints
+	// partitions specify a type.
+	// The default type is also used for the automatically created /boot
+	// filesystem if it is a supported type for that fileystem. If it is not,
+	// xfs is used as a fallback.
+	DefaultFSType FSType
+
+	// RequiredMinSizes defines a map of minimum sizes for specific
+	// directories. These indirectly control the minimum sizes of partitions. A
+	// directory with a required size will set the minimum size of the
+	// partition with the mountpoint that contains the directory. Additional
+	// directory requirements are additive, meaning the minimum size for a
+	// mountpoint's partition is the sum of all the required directory sizes it
+	// will contain.
+	RequiredMinSizes map[string]uint64
+}
+
+// Returns the default filesystem type if the fstype is empty. If both are
+// empty/none, returns an error.
+func (options *CustomPartitionTableOptions) getfstype(fstype string) (string, error) {
+	if fstype != "" {
+		return fstype, nil
+	}
+
+	if options.DefaultFSType == FS_NONE {
+		return "", fmt.Errorf("no filesystem type defined and no default set")
+	}
+
+	return options.DefaultFSType.String(), nil
+}
+
+// NewCustomPartitionTable creates a partition table based almost entirely on the partitioning customizations from a blueprint.
+func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization, options *CustomPartitionTableOptions, rng *rand.Rand) (*PartitionTable, error) {
+	if options == nil {
+		// init options with empty struct to use defaults
+		options = &CustomPartitionTableOptions{
+			PartitionTableType: PT_GPT,
+		}
+	}
+
+	if customizations == nil {
+		customizations = &blueprint.PartitioningCustomization{}
+	}
+
+	// validate the partitioning customizations before using them
+	if err := customizations.Validate(); err != nil {
+		return nil, fmt.Errorf("error generating partition table: %w", err)
+	}
+
+	pt := &PartitionTable{}
+
+	switch options.PartitionTableType {
+	case PT_GPT, PT_DOS:
+		pt.Type = options.PartitionTableType.String()
+	case PT_NONE:
+		// default to "gpt"
+		pt.Type = PT_GPT.String()
+	default:
+		return nil, fmt.Errorf("error generating partition table: invalid partition table type enum value: %d", options.PartitionTableType)
+	}
+
+	// The boot type will be the default only if it's a supported filesystem
+	// type for /boot (ext4 or xfs). Otherwise, we default to xfs.
+	// FS_NONE also falls back to xfs.
+	var bootType FSType
+	switch options.DefaultFSType {
```

Comment:
(nitpick) maybe move this closer to line 276 where it is used? Right now it looks like it's related to `addPartitionsForBootMode()` but for that it's not actually needed (i.e. something like):
```go
diff --git a/pkg/disk/partition_table.go b/pkg/disk/partition_table.go
index 8a682668e..31de9f783 100644
--- a/pkg/disk/partition_table.go
+++ b/pkg/disk/partition_table.go
@@ -262,18 +262,19 @@ func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization
        // The boot type will be the default only if it's a supported filesystem
        // type for /boot (ext4 or xfs). Otherwise, we default to xfs.
        // FS_NONE also falls back to xfs.
-       var bootType FSType
-       switch options.DefaultFSType {
-       case FS_EXT4, FS_XFS:
-               bootType = options.DefaultFSType
-       default:
-               bootType = FS_XFS
-       }
 
        if err := addPartitionsForBootMode(pt, options.BootMode); err != nil {
                return nil, fmt.Errorf("error generating partition table: %w", err)
        }
+
        if customizations.LVM != nil || customizations.Btrfs != nil {
+               var bootType FSType
+               switch options.DefaultFSType {
+               case FS_EXT4, FS_XFS:
+                       bootType = options.DefaultFSType
+               default:
+                       bootType = FS_XFS
+               }
                // we need a /boot partition to boot LVM or Btrfs, create boot
                // partition if it does not already exist
                ensureBoot(pt, bootType)

```

Reply:
Done